packages/web3-providers-ws/src/index.ts

```diff
@@ -0,0 +1,391 @@
+
+	private reconnectAttempts!: number;
+	private lastDataChunk!: string;
+	private lastChunkTimeout!: NodeJS.Timeout;
+
+	public constructor(clientUrl: string, wsProviderOptions?: WebSocketOptions) {
+		super();
+		if (!WebSocketProvider.validateProviderUrl(clientUrl))
+			throw new InvalidClientError(clientUrl);
+
+		this.clientUrl = clientUrl;
+		if (wsProviderOptions !== undefined) this.wsProviderOptions = wsProviderOptions;
+		else
+			this.wsProviderOptions = {
+				customTimeout: 1000 * 15,
+			};
+
+		if (!this.wsProviderOptions.customTimeout) this.wsProviderOptions.customTimeout = 1000 * 15;
+
+		if (!this.wsProviderOptions.reconnectOptions)
+			this.wsProviderOptions.reconnectOptions = {
+				auto: false,
+				delay: 5000,
+				maxAttempts: 5,
+				onTimeout: false,
+			};
+
+		this.requestQueue = new Map<JsonRpcId, WSRequestItem>();
+		this.sentQueue = new Map<JsonRpcId, WSRequestItem>();
+
+		this.providerStatus = 'disconnected';
+
+		this.init();
+		this.connect();
+	}
+
+	private static validateProviderUrl(providerUrl: string): boolean {
+		return typeof providerUrl === 'string' ? /^ws(s)?:\/\//i.test(providerUrl) : false;
+	}
+
+	public getStatus(): Web3BaseProviderStatus {
+		return this.providerStatus;
+	}
+
+	/* eslint-disable class-methods-use-this */
+	public supportsSubscriptions(): boolean {
+		return true;
+	}
+
+	public on<T = JsonRpcResult>(
+		type: 'message' | string,
+		callback: Web3BaseProviderCallback<T>,
+	): void {
+		this.wsEventEmitter.on(type, callback);
+	}
+
+	public once<T = JsonRpcResult>(type: string, callback: Web3BaseProviderCallback<T>): void {
+		this.wsEventEmitter.once(type, callback);
+	}
+
+	public removeListener(type: string, callback: Web3BaseProviderCallback): void {
+		this.wsEventEmitter.removeListener(type, callback);
+	}
+
+	public connect(): void {
+		try {
+			this.providerStatus = 'connecting';
+
+			this.webSocketConnection = new W3WS(
+				this.clientUrl,
+				this.wsProviderOptions?.protocol,
+				undefined,
+				this.wsProviderOptions?.headers,
+				this.wsProviderOptions?.requestOptions,
+				this.wsProviderOptions?.clientConfig,
+			);
+
+			this.addSocketListeners();
+		} catch (e) {
+			this.providerStatus = 'disconnected';
+			throw new InvalidConnectionError(this.clientUrl);
+		}
+	}
+
+	public disconnect(code: number, reason: string): void {
+		this.removeSocketListeners();
+		this.webSocketConnection?.close(code || 1000, reason);
+		this.providerStatus = 'disconnected';
+	}
+
+	public reset(): void {
+		this.sentQueue.clear();
+		this.requestQueue.clear();
+
+		this.init();
+		this.removeSocketListeners();
+		this.addSocketListeners();
+	}
+
+	public async request<T = JsonRpcResponse, T2 = unknown[]>(
+		request: JsonRpcPayload<T2>,
+	): Promise<T> {
+		if (this.webSocketConnection === undefined)
+			throw new Error('WebSocket connection is undefined');
+
+		if (request.id === undefined) throw new Error('Request Id not defined');
```

Comment:
The input of function is the request, and request contains only `method` and `parameters`. The `id` and other fields are part of the payload and which are optional in the parameters. The transport had to sent the payload and for that it needed to be created and validated before sending. So if user didn't provided the `id` it's not an issue and provider should handle creating the payload internally.

Reply:
GRASP High cohesion, transport should only focus its function 'transport', instead of other things: managing and modifying param obj internals.

Reply:
The http provider [generates an ID](https://github.com/ChainSafe/web3.js/blob/4.x/packages/web3-providers-http/src/index.ts#L55) if it's not provided by the user, and I think WS and IPC providers should do the same. However, I can't find how IDs are generated in `1.x`

Reply:
I think generating id from http provider should also be removed, as it will add extra collision check responsibility in all providers: 
- managing internal collisions ( if id is generated inside provider ) 
- managing external collisions of user's passed requests with internal generated ones